File: calculator.js
// Calculator State
const PI = Math.acos(-1);
const E = Math.exp(1);

let calculator = {
    memory: 0,
    ans: 0,
    history: [],
    display: '',
};

// DOM Elements
const displayInput = document.getElementById('display');
const resultDisplay = document.getElementById('result');
const historyList = document.getElementById('historyList');
const memoryDisplay = document.getElementById('memoryDisplay');
const historyPanel = document.getElementById('historyPanel');

// Load from localStorage on startup
function loadState() {
    const saved = localStorage.getItem('calculator-state');
    if (saved) {
        const state = JSON.parse(saved);
        calculator.memory = state.memory || 0;
        calculator.ans = state.ans || 0;
        calculator.history = state.history || [];
        updateMemoryDisplay();
        updateHistoryDisplay();
    }
}

// Save to localStorage
function saveState() {
    localStorage.setItem('calculator-state', JSON.stringify(calculator));
}

// Update displays
function updateDisplay() {
    displayInput.value = calculator.display;
}

function updateMemoryDisplay() {
    memoryDisplay.textContent = formatNumber(calculator.memory);
    saveState();
}

function updateHistoryDisplay() {
    historyList.innerHTML = '';
    calculator.history.slice().reverse().forEach((entry, index) => {
        const div = document.createElement('div');
        div.className = 'history-item';
        div.innerHTML = `<span>${index + 1}.</span> ${escapeHtml(entry)}`;
        div.onclick = () => {
            // Extract just the expression part
            const parts = entry.split(' = ');
            if (parts[0]) {
                calculator.display = parts[0];
                updateDisplay();
            }
        };
        historyList.appendChild(div);
    });
    saveState();
}

function escapeHtml(text) {
    const div = document.createElement('div');
    div.textContent = text;
    return div.innerHTML;
}

// Append character or function
function appendChar(char) {
    if (char === 'pi') {
        calculator.display += 'pi';
    } else if (char === 'e') {
        calculator.display += 'e';
    } else if (char === 'ans') {
        calculator.display += 'ans';
    } else {
        calculator.display += char;
    }
    updateDisplay();
}

function appendFunction(func) {
    calculator.display += func + '(';
    updateDisplay();
}

// Clear and backspace
function clearDisplay() {
    calculator.display = '';
    updateDisplay();
    resultDisplay.textContent = '0';
}

function backspace() {
    calculator.display = calculator.display.slice(0, -1);
    updateDisplay();
}

// Memory operations
function storeMemory() {
    if (resultDisplay.textContent !== '0') {
        calculator.memory = calculator.ans;
        updateMemoryDisplay();
        showNotification('Stored memory');
    }
}

function recallMemory() {
    calculator.display += calculator.memory.toString();
    updateDisplay();
}

function clearMemory() {
    calculator.memory = 0;
    updateMemoryDisplay();
    showNotification('Memory cleared');
}

// History
function clearHistory() {
    calculator.history = [];
    updateHistoryDisplay();
    showNotification('History cleared');
}

function toggleHistory() {
    historyPanel.classList.toggle('hidden');
}

// Parser - Port of C++ calculator logic
class Parser {
    constructor(expr) {
        this.expr = expr;
        this.pos = 0;
    }

    skipWS() {
        while (this.pos < this.expr.length && /\s/.test(this.expr[this.pos])) {
            this.pos++;
        }
    }

    peek() {
        this.skipWS();
        return this.pos < this.expr.length ? this.expr[this.pos] : '\0';
    }

    consume() {
        this.skipWS();
        return this.pos < this.expr.length ? this.expr[this.pos++] : '\0';
    }

    match(c) {
        if (this.peek() === c) {
            this.pos++;
            return true;
        }
        return false;
    }

    parseExpr() {
        let result = this.parseTerm();
        while (true) {
            if (this.match('+')) {
                result += this.parseTerm();
            } else if (this.match('-')) {
                result -= this.parseTerm();
            } else {
                break;
            }
        }
        return result;
    }

    parseTerm() {
        let result = this.parseFactor();
        while (true) {
            if (this.match('*')) {
                result *= this.parseFactor();
            } else if (this.match('/')) {
                const d = this.parseFactor();
                if (d === 0) throw new Error('Division by zero');
                result /= d;
            } else if (this.match('%')) {
                const d = this.parseFactor();
                if (d === 0) throw new Error('Modulo by zero');
                result %= d;
            } else {
                break;
            }
        }
        return result;
    }

    parseFactor() {
        let base = this.parseBase();
        if (this.match('^')) {
            let exp = this.parseFactor();
            return Math.pow(base, exp);
        }
        return base;
    }

    parseBase() {
        if (this.peek() === '-') {
            this.consume();
            return -this.parsePrimary();
        }
        if (this.peek() === '+') {
            this.consume();
            return this.parsePrimary();
        }
        return this.parsePrimary();
    }

    parsePrimary() {
        this.skipWS();
        if (this.peek() === '(') {
            this.consume();
            const val = this.parseExpr();
            if (!this.match(')')) throw new Error("Missing closing ')'");
            return val;
        }
        if (/\d/.test(this.peek()) || this.peek() === '.') {
            return this.parseNumber();
        }
        if (/[a-zA-Z_]/.test(this.peek())) {
            return this.parseIdent();
        }
        throw new Error(`Unexpected character: '${this.peek()}'`);
    }

    parseNumber() {
        const start = this.pos;
        while (this.pos < this.expr.length && (/\d/.test(this.expr[this.pos]) || this.expr[this.pos] === '.')) {
            this.pos++;
        }
        if (this.pos < this.expr.length && /[eE]/.test(this.expr[this.pos])) {
            this.pos++;
            if (this.pos < this.expr.length && /[+-]/.test(this.expr[this.pos])) this.pos++;
            while (this.pos < this.expr.length && /\d/.test(this.expr[this.pos])) this.pos++;
        }
        const tok = this.expr.substring(start, this.pos);
        const num = parseFloat(tok);
        if (isNaN(num)) throw new Error(`Invalid number: ${tok}`);
        return num;
    }

    parseIdent() {
        const start = this.pos;
        while (this.pos < this.expr.length && /[a-zA-Z0-9_]/.test(this.expr[this.pos])) {
            this.pos++;
        }
        const name = this.expr.substring(start, this.pos).toLowerCase();

        if (name === 'pi' || name === 'π') return PI;
        if (name === 'e') return E;
        if (name === 'ans') return calculator.ans;

        this.skipWS();
        if (this.peek() !== '(') {
            throw new Error(`Unknown identifier: ${name}`);
        }
        this.consume();
        const a = this.parseExpr();

        // Two-argument functions
        if (name === 'pow') {
            if (!this.match(',')) throw new Error('pow needs two arguments');
            const b = this.parseExpr();
            if (!this.match(')')) throw new Error("Missing ')'");
            return Math.pow(a, b);
        }

        if (name === 'hypot') {
            if (!this.match(',')) throw new Error('hypot needs two arguments');
            const b = this.parseExpr();
            if (!this.match(')')) throw new Error("Missing ')'");
            return Math.hypot(a, b);
        }

        if (name === 'log' && this.peek() === ',') {
            this.consume();
            const base = this.parseExpr();
            if (!this.match(')')) throw new Error("Missing ')'");
            if (base <= 0 || base === 1) throw new Error('Invalid logarithm base');
            return Math.log(a) / Math.log(base);
        }

        // One-argument functions
        if (!this.match(')')) throw new Error("Missing ')'");

        switch (name) {
            case 'sin': return Math.sin(a);
            case 'cos': return Math.cos(a);
            case 'tan': return Math.tan(a);
            case 'asin':
                if (a < -1 || a > 1) throw new Error('asin domain error');
                return Math.asin(a);
            case 'acos':
                if (a < -1 || a > 1) throw new Error('acos domain error');
                return Math.acos(a);
            case 'atan': return Math.atan(a);
            case 'sinh': return Math.sinh(a);
            case 'cosh': return Math.cosh(a);
            case 'tanh': return Math.tanh(a);
            case 'sqrt':
                if (a < 0) throw new Error('sqrt of negative');
                return Math.sqrt(a);
            case 'cbrt': return Math.cbrt(a);
            case 'log':
                if (a <= 0) throw new Error('log of non-positive');
                return Math.log(a);
            case 'log2':
                if (a <= 0) throw new Error('log2 of non-positive');
                return Math.log2(a);
            case 'log10':
                if (a <= 0) throw new Error('log10 of non-positive');
                return Math.log10(a);
            case 'exp': return Math.exp(a);
            case 'abs': return Math.abs(a);
            case 'ceil': return Math.ceil(a);
            case 'floor': return Math.floor(a);
            case 'round': return Math.round(a);
            case 'deg': return a * 180.0 / PI;
            case 'rad': return a * PI / 180.0;
            case 'fact':
                if (a < 0 || a !== Math.floor(a)) {
                    throw new Error('Factorial only for non-negative integers');
                }
                let r = 1;
                for (let i = 2; i <= a; i++) r *= i;
                return r;
            default:
                throw new Error(`Unknown function: ${name}`);
        }
    }

    parse() {
        const result = this.parseExpr();
        this.skipWS();
        if (this.pos !== this.expr.length) {
            throw new Error(`Unexpected token near: ${this.expr.substring(this.pos)}`);
        }
        return result;
    }
}

// Calculate
function calculate() {
    if (!calculator.display) {
        return;
    }

    try {
        const processed = calculator.display.replace(/mem/g, calculator.memory.toString());
        const parser = new Parser(processed);
        const result = parser.parse();

        calculator.ans = result;
        resultDisplay.textContent = formatNumber(result);

        const entry = `${calculator.display} = ${formatNumber(result)}`;
        calculator.history.push(entry);
        if (calculator.history.length > 50) {
            calculator.history.shift();
        }
        updateHistoryDisplay();
        saveState();
    } catch (error) {
        resultDisplay.textContent = `Error: ${error.message}`;
        showNotification(`Error: ${error.message}`, 'error');
    }
}

// Format number output
function formatNumber(v) {
    if (!isFinite(v)) {
        return v > 0 ? 'Infinity' : '-Infinity';
    }
    if (isNaN(v)) {
        return 'NaN';
    }
    if (v === Math.floor(v) && Math.abs(v) < 1e15) {
        return Math.floor(v).toString();
    }
    return v.toPrecision(10).replace(/\.?0+$/, '');
}

// Notification
function showNotification(msg, type = 'info') {
    const notification = document.createElement('div');
    notification.style.cssText = `
        position: fixed;
        top: 20px;
        right: 20px;
        padding: 12px 20px;
        background: ${type === 'error' ? '#f44336' : '#4caf50'};
        color: white;
        border-radius: 8px;
        z-index: 1000;
        animation: slideIn 0.3s ease;
    `;
    notification.textContent = msg;
    document.body.appendChild(notification);
    setTimeout(() => notification.remove(), 2000);
}

// Add keyboard support
document.addEventListener('keydown', (e) => {
    if (e.key === 'Enter') {
        e.preventDefault();
        calculate();
    } else if (e.key === 'Escape') {
        clearDisplay();
    } else if (e.key === 'Backspace') {
        backspace();
    } else if (e.key === '+' || e.key === '-' || e.key === '*' || e.key === '/' || e.key === '%' || e.key === '^') {
        e.preventDefault();
        appendChar(e.key);
    } else if (/\d|\.|\(|\)|,/.test(e.key)) {
        e.preventDefault();
        appendChar(e.key);
    }
});

// Initialize
loadState();
updateDisplay();

File: Calculator.cpp
#include<bits/stdc++.h>
using namespace std;
const double PI = acos(-1.0);
const double E  = exp(1.0);
struct 
Calculator {
    double memory = 0.0;
    double ans    = 0.0;
    vector<string> history;
    string expr;
    size_t pos = 0;
    void skipWS() {
        while (pos<expr.size() && isspace((unsigned char)expr[pos])) ++pos;
    }

    char peek() {
        skipWS();
        return (pos < expr.size()) ? expr[pos] : '\0';
    }

    char consume() {
        skipWS();
        return (pos < expr.size()) ? expr[pos++] : '\0';
    }

    bool match(char c) {
        if (peek() == c) { ++pos; return true; }
        return false;
    }
    double parseExpr() {
        double result = parseTerm();
        while (true) {
            if      (match('+')) result += parseTerm();
            else if (match('-')) result -= parseTerm();
            else break;
        }
        return result;
    }

    double parseTerm() {
        double result = parseFactor();
        while (true) {
            if (match('*')) {
                result *= parseFactor();
            } else if (match('/')) {
                double d = parseFactor();
                if (d == 0.0) throw std::runtime_error("Division by zero");
                result /= d;
            } else if (match('%')) {
                double d = parseFactor();
                if (d == 0.0) throw std::runtime_error("Modulo by zero");
                result = fmod(result, d);
            } else break;
        }
        return result;
    }
    double parseFactor() {
        double base = parseBase();
        if (match('^')) {
            double exp = parseFactor();
            return pow(base, exp);
        }
        return base;
    }
    double parseBase() {
        if (peek() == '-') { consume(); return -parsePrimary(); }
        if (peek() == '+') { consume(); return  parsePrimary(); }
        return parsePrimary();
    }
    double parsePrimary() {
        skipWS();
        if (peek() == '(') {
            consume();
            double val = parseExpr();
            if (!match(')')) throw std::runtime_error("Missing closing ')'");
            return val;
        }
        if (isdigit((unsigned char)peek()) || peek() == '.') {
            return parseNumber();
        }
        if (isalpha((unsigned char)peek()) || peek() == '_') {
            return parseIdent();
        }

        throw runtime_error(
            string("Unexpected character: '") + peek() + "'");
    }

    double parseNumber() {
        size_t start = pos;
        while (pos<expr.size() && (isdigit((unsigned char)expr[pos]) || expr[pos] == '.')) ++pos;
        if (pos < expr.size() && (expr[pos] == 'e' || expr[pos] == 'E')) {
            ++pos;
            if (pos < expr.size() && (expr[pos] == '+' || expr[pos] == '-')) ++pos;
            while (pos < expr.size() && isdigit((unsigned char)expr[pos])) ++pos;
        }
        string tok = expr.substr(start, pos - start);
        try { return stod(tok); }
        catch (...) { throw runtime_error("Invalid number: " + tok); }
    }

    double parseIdent() {
        size_t start = pos;
        while (pos < expr.size() &&
               (isalnum((unsigned char)expr[pos]) || expr[pos] == '_')) ++pos;
        string name = expr.substr(start, pos - start);
        if (name == "pi" || name == "PI")  return PI;
        if (name == "e"  || name == "E")   return E;
        if (name == "ans")                 return ans;
        skipWS();
        if (peek() != '(')
            throw runtime_error("Unknown identifier: " + name);
        consume(); // '('
        double a = parseExpr();
        if (name == "pow") {
            if (!match(',')) throw runtime_error("pow needs two arguments");
            double b = parseExpr();
            if (!match(')')) throw runtime_error("Missing ')'");
            return pow(a, b);
        }
        if (name == "hypot") {
            if (!match(',')) throw runtime_error("hypot needs two arguments");
            double b = parseExpr();
            if (!match(')')) throw runtime_error("Missing ')'");
            return hypot(a, b);
        }
        if (name == "log" && peek() == ',') {
            consume(); 
            double base = parseExpr();
            if (!match(')')) throw runtime_error("Missing ')'");
            if (base <= 0 || base == 1)
                throw runtime_error("Invalid logarithm base");
            return log(a) / log(base);
        }

        if (!match(')')) throw runtime_error("Missing ')'");
        if (name == "sin")   return sin(a);
        if (name == "cos")   return cos(a);
        if (name == "tan")   return tan(a);
        if (name == "asin")  { if (a<-1||a>1) throw runtime_error("asin domain error"); return asin(a); }
        if (name == "acos")  { if (a<-1||a>1) throw runtime_error("acos domain error"); return acos(a); }
        if (name == "atan")  return atan(a);
        if (name == "sinh")  return sinh(a);
        if (name == "cosh")  return cosh(a);
        if (name == "tanh")  return tanh(a);
        if (name == "sqrt")  { if (a<0) throw runtime_error("sqrt of negative"); return sqrt(a); }
        if (name == "cbrt")  return cbrt(a);
        if (name == "log"  ) { if (a<=0) throw runtime_error("log of non-positive"); return log(a); }
        if (name == "log2" ) { if (a<=0) throw runtime_error("log2 of non-positive"); return log2(a); }
        if (name == "log10") { if (a<=0) throw runtime_error("log10 of non-positive"); return log10(a); }
        if (name == "exp"  ) return exp(a);
        if (name == "abs"  ) return fabs(a);
        if (name == "ceil" ) return ceil(a);
        if (name == "floor") return floor(a);
        if (name == "round") return round(a);
        if (name == "deg"  ) return a * 180.0 / PI;
        if (name == "rad"  ) return a * PI / 180.0;
        if (name == "fact" ) {
            if (a < 0 || a != floor(a))
                throw runtime_error("Factorial only for non-negative integers");
            long long n = (long long)a, r = 1;
            for (long long i = 2; i <= n; ++i) r *= i;
            return (double)r;
        }

        throw runtime_error("Unknown function: " + name);
    }


    double evaluate(const std::string& input) {
        expr = input;
        pos  = 0;
        double result = parseExpr();
        skipWS();
        if (pos != expr.size())
            throw runtime_error("Unexpected token near: " + expr.substr(pos));
        ans = result;
        return result;
    }
};
string format(double v) {
    if (isnan(v))  return "NaN";
    if (isinf(v))  return v > 0 ? "Infinity" : "-Infinity";
    if (v == floor(v) && fabs(v) < 1e15)
        return to_string((long long)v);
    ostringstream ss;
    ss << setprecision(10) << v;
    return ss.str();
}
void printHelp() {
    cout << R"(
               Advanced C++ Calculator                    
  Operators : + - * / % ^  (^ = power, right-assoc)       
  Constants : pi, e, ans                                
                                                          
 Functions (one-arg):                                    
    sin(x)  cos(x)  tan(x)   asin(x)  acos(x)  atan(x)        
    sinh(x) cosh(x) tanh(x)  sqrt(x)  cbrt(x)  exp(x)     
    log(x)  log2(x) log10(x) abs(x)   ceil(x)  floor(x)   
    round(x) deg(x) rad(x)   fact(x)                      
                                                          
  Functions (two-arg):                                   
    pow(x,y)   hypot(x,y)   log(x, base)                  
                                                          
  Commands:                                               
    store      save last result to memory                
    recall     retrieve memory value                     
    mclear     clear memory                              
    history    show calculation history                  
    clear      clear history                             
    help       show this help                            
    exit/quit  exit the calculator                       
)";
}

int main() {
    Calculator calc;
    string line;

    printHelp();
    cout << "\nType an expression and press Enter.\n\n";

    while (true) {
        cout << "calc> ";
        if (!getline(cin,line)) break;
        auto ltrim=line.find_first_not_of("\t\r\n");
        if (ltrim==string::npos) continue;
        line=line.substr(ltrim);
        auto rtrim=line.find_last_not_of("\t\r\n");
        if (rtrim != string::npos)line=line.substr(0,rtrim+1);
        if (line.empty()) continue;
        string low=line;
        transform(low.begin(),low.end(),low.begin(),::tolower);
        if(low=="exit"||low=="quit"){
            cout << "Goodbye!\n";
            break; 
        }
        if (low=="help") {printHelp(); continue;}
        if (low == "store") {
            calc.memory = calc.ans;
            cout << "  Stored " << format(calc.memory) << " in memory.\n";
            continue;
        }
        if (low == "recall") {
            cout << "  Memory = " << format(calc.memory) << "\n";
            continue;
        }
        if (low == "mclear") {
            calc.memory = 0;
            cout << "  Memory cleared.\n";
            continue;
        }
        if (low == "history") {
            if (calc.history.empty()) {
                cout << "  (no history yet)\n";
            } else {
                for (size_t i = 0; i < calc.history.size(); ++i)
                    cout << "  [" << (i+1) << "] " << calc.history[i] << "\n";
            }
            continue;
        }
        if (low == "clear") {
            calc.history.clear();
            cout << "  History cleared.\n";
            continue;
        }
        std::string processed = line;
        size_t mpos;
        while ((mpos = processed.find("mem")) != string::npos)
            processed.replace(mpos, 3, std::to_string(calc.memory));
        try {
            double result = calc.evaluate(processed);
            string entry = line + " = " + format(result);
            calc.history.push_back(entry);
            cout << "  = " << format(result) << "\n";
        } catch (const std::exception& ex) {
            std::cout << "  Error: " << ex.what() << "\n";
        }
    }

    return 0;
}